### ISSUE TITLE
Memory and threading (OOM, OutOfDirectMemoryError, duplicates)

### ISSUE THREAD

**Issue by dariocazas** [GoogleCloudPlatform/pubsub#279]
My team is working in a streaming solution with a part on-premise and part on the cloud. In the on-prem layer, we using Kafka and in cloud part using GCP PubSub.

## Empirical - part 1 (OOM)

Actually, we do a load test using this connect to upload data from on-prem Kafka to GCP PubSub, using a single topic with one partition and 120k events (1GB of snappy compressed data). The config using for this connector plugin (v0.8) is:

```json
{
    "connector.class": "com.google.pubsub.kafka.sink.CloudPubSubSinkConnector",
    "name": "sink_gcp-pubsub_topic1",
    "topics": "topic1",
    "tasks.max": "1",
    "key.converter.schemas.enable": "false",
    "value.converter.schemas.enable": "false",
    "value.converter": "org.apache.kafka.connect.converters.ByteArrayConverter",
    "key.converter": "org.apache.kafka.connect.converters.ByteArrayConverter",
    "cps.project": "my-gcp-project",
    "gcp.credentials.file.path": "/path/to/secrect.json",
    "cps.topic": "topic1",
    "headers.publish": "true",
    "metadata.publish": "true"
}
```

The Kafka connect runs in one worker node with 24GB of RAM. If we configure the JVM heap with 12GB, in less than a minute you get out of memory and the system kills the JVM. Recap:
* One task
* 1GB of compressed data with snappy
* Kafka connect worker with 24GB of RAM gets an out of memory

If we lower the heap size to 8GB, we see traces that indicate containment:
```
java.util.concurrent.ExecutionException: io.grpc.netty.shaded.io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 2097152 byte(s) of direct memory (used: 8587837719, max: 8589934592)
	at com.google.pubsub.kafka.sink.CloudPubSubSinkTask.flush(CloudPubSubSinkTask.java:355)
	at org.apache.kafka.connect.sink.SinkTask.preCommit(SinkTask.java:125)
```

In this case, the max memory used is about 20GB-21GB, and we can create other connects that sink to PubSub without OOM errors. 

## Empirical - part 2 (retries and duplicates)

The OutOfDirectMemoryError causes fail to commit in Kafka (although the events are published to PubSub). With the previous configuration and based on retries, the previous upload ends in 1 hour (in different tests, it took from 40min to 1h30min) uploading millions (many duplicates from original 120k). 

We have tested behavior by lowering heap to 5GB to simulate cases in which other connects race for the resources of the machine. In this case, the upload never ends (the commit always fails).
We test changing these properties in several tests without success:
- maxBufferSize increase to generate bigger batches (try to reduce the number of requests)
- maxBufferBytes cannot be increased
- maxDelayThresholdMs to 1000 (same reason like maxBufferSize)

Next target: review global properties (worker properties)
- [offset.flush.interval.ms](https://kafka.apache.org/documentation/#connectconfigs_offset.flush.interval.ms) default value 60sg
- [offset.flush.timeout.ms](https://kafka.apache.org/documentation/#connectconfigs_offset.flush.timeout.ms) default value 5sg

The first test, using interval=12sg and gets commit timeout in Kafka. Change the timeout to 15sg to reproduce the OutOfDirectMemoryError.
Next test, **using interval=5sg, the upload runs fine without any error and upload all the data**.

## Empirical - part 3 (threads)

For each connector created for each test, we see that around 100 threads are generated, which are only destroyed once the connector is removed.

### Memory usage

In this test, the connector plugin consumes twice the size of the heap it has available, and only with one task moving 1GB of compressed data to GCP saturates a 24GB worker node. It is the expected behavior? Maybe other points that we need to consider? 

### Containment

Changing the global property of the worker that forces flush the connects from 60sg to 5sg solves this problem, but IMHO is the work of the connector manages your memory without this force solution. An autoflush when several configurable margins of queue batches open the door to managing it manually by configuration. What is your opinion? How to solve the containment without change a global property that affects all connectors?

### Thread management

The number of threads generated for each sink operation is very high, and we need to create one sink connector for each topic to replicate to PubSub. With a few topics to replicate we need thousands of threads, which raises a scaling problem for us. Maybe is possible to configure an executor size or similar option with a minor size?

Other solutions like MirrorMaker allow you to configure several topics to replicate in one connector, and with this option, we could group topics that share that high volume of threads (from the ignorance of why so many threads are currently being raised)

**Comment by jordanbull** [comment]
For parts 1 and 2, we were seeing the same OODM behavior and submitted #282 to deal with it. We've been running that change in production for about a month now and the OODM issue has not recurred since.

**Comment by samarthsingal** [comment]
For part 3, #288 changes to using the same threadpool for all Publisher/Subscriber objects across tasks. This should reduce the number of threads seen with multiple topics.

Both changes are released in [v0.10-alpha](https://github.com/GoogleCloudPlatform/pubsub/releases/tag/v0.10-alpha).

**Comment by dariocazas** [comment]
Hi. Sorry for my delayed feedback. Testing the new connector we can see:

* The new thread management works fine, and as side effect we can see a reduction in the memory footprint
* The new configuration property `maxOutstandingRequestBytes` works fine to let us containment (memory and network usage)

However, we can see that if the output network of the cluster is exceeded, the sink connectors upload several duplicates in PubSub. In cases of events with high compression in Kafka, we can see a x5-x10 of outgoing bandwich to PubSub. Solutions like the issue #280 help to us to avoid issues like high duplicates in upload to PubSub on these cases.

Thank you very much for your work, is very usefull to us!
